src/main/scala/optics/poly/GenLens.scala

```diff
@@ -77,15 +82,146 @@ object GenLens {
+        case Inlined(None, Nil, lambda @ Lambda(params, body)) => Some((lambda))
+        case _ => None
+      }
+    }
+
+    object Path {
+
+      private def recur(tree: Term, selects: List[String]): Option[(List[String], Term)] = tree match {
+        case Select(qual, name) => recur(qual, name :: selects)
+        // stop extraction at this stage
+        case expr if selects.nonEmpty => Some((selects, expr))
+        case _ => None
+      }
+
+      def unapply(t: Term): Option[(List[String], Term)] = recur(t, Nil)
+    }
+
+    object Operator {
+      def unapply(t: Term): Option[(Ident, List[TypeTree], Term, Term)] = t match {
+        case Apply(Apply(TypeApply(ident @ Ident(_), tpe), rhs :: Nil), witness :: Nil) =>
+          Some((ident, tpe, witness, rhs))
+        case _ => None
+      }
+    }
+
+    case class State(from: List[Type], to: List[Type])
+
+    def genLens(from: Type, to: Type, term: Term): Term = {
+      val ft = from.seal.asInstanceOf[scala.quoted.Type[_ <: Any]]
+      val tt = to.seal.asInstanceOf[scala.quoted.Type[_ <: Any]]
+      val tpe = '[$ft => $tt]
+      '{ GenLens.uncurried(${term.seal.cast(using tpe)}) }.unseal
+    }
+
+    def selector(from: Type, to: Type, sel: List[String]): Term = {
+
+      def body(term: Term, sel: List[String]): Term = {
+        sel match {
+          case Nil => term
+          case x :: xs =>
+            body(Select.unique(term, x), xs)
+        }
+      }
+
+      Lambda(
+        MethodType(List("x"))(const(List(from)), const(to)),
+        args => args match {
+          case arg :: Nil => body(Ref(arg.symbol), sel)
+          // shouldn't ever go down this branch
+          case _ => ???
+        }
+      )
+    }
+
+    def fieldType(sel: List[String], t: Type): Type =
+      sel match {
+        case Nil => t
+        case x :: xs =>
+          val ValDef(field, typeTree, _) = t.classSymbol.get.field(x).tree
+          fieldType(xs, typeTree.tpe)
+      }
+
+    def fold(term: Term)(state: State): (State, Term) =
+      term match {
+        case Function(Lambda(_, body)) =>
+          fold(body)(state)
+        case Path(sel, Ident(_)) =>
+          state match {
+            case State(from :: Nil, _) =>
+              val to = fieldType(sel, from)
+              (
+                state.copy(from = to :: state.from),
+                genLens(
+                  from,
+                  to,
+                  selector(from, to, sel)
+                )
+              )
+            case _ => ???
+          }
+        case Path(sel, Operator(Ident("?"), _ :: err :: success :: Nil, witness, rhs)) =>
+          fold(rhs)(state) match {
+            case (State(from :: _,  _), term) =>
+              fieldType(sel, success.tpe)
+              val errt = err.tpe.seal.asInstanceOf[scala.quoted.Type[_ <: Any]]
+              val successt = success.tpe.seal.asInstanceOf[scala.quoted.Type[_ <: Any]]
+              (
+                state,
+                Apply(
```

Comment:
Here's where the magic happens! We're calling `composeIso` with the witness we grabbed in the `dsl`.